pallets/msa/src/lib.rs

```diff
@@ -883,6 +976,90 @@ impl<T: Config> Pallet<T> {
 		}
 		Ok(Some(schemas.0.into()))
 	}
+
+	/// Add a signature to the PayloadSignatureRegistry based on a virtual "bucket" grouping.
+	/// First reset the virtual bucket if necessary, then add the new entry.
+	pub fn register_signature(
+		signature: &MultiSignature,
+		current_block: T::BlockNumber,
+		mortality_block: T::BlockNumber,
+	) -> DispatchResult {
+		if Self::mortality_block_limit(current_block) < mortality_block {
+			Err(Error::<T>::MortalityTooHigh.into())
+		} else {
+			let bucket_num = Self::bucket_for(mortality_block.into());
+			<PayloadSignatureRegistry<T>>::try_mutate(
+				bucket_num,
+				signature,
+				|maybe_mortality_block| -> DispatchResult {
+					ensure!(maybe_mortality_block.is_none(), Error::<T>::BucketKeyExists);
+
+					Self::reset_virtual_bucket_if_needed(mortality_block, bucket_num)?;
```

Comment:
It does seem like you can also do this `on_initialize`. Having this process here will cause some transactions to have higher weight which doesn't seem to be fair or consistent

Reply:
Good point